src/qibolab/instruments/rfsoc.py

```diff
@@ -1,54 +1,164 @@
 """ RFSoC FPGA driver.
 
-This driver needs the library Qick installed
-Supports the following FPGA:
+This driver needs a Qibosoq server installed and running
+Tested on the following FPGA:
  *   RFSoC 4x2
+ *   ZCU111
 """
 
-import pickle
+import json
 import socket
-from dataclasses import dataclass
-from typing import Dict, List, Optional, Tuple, Union
+from dataclasses import asdict
+from typing import Dict, List, Tuple, Union
 
 import numpy as np
+import qibosoq.components as rfsoc
 
 from qibolab import AcquisitionType, AveragingMode, ExecutionParameters
 from qibolab.instruments.abstract import AbstractInstrument
 from qibolab.platform import Qubit
-from qibolab.pulses import PulseSequence, PulseType
+from qibolab.pulses import Drag, Gaussian, Pulse, PulseSequence, PulseType, Rectangular
 from qibolab.result import IntegratedResults, SampleResults
 from qibolab.sweeper import Parameter, Sweeper
 
+HZ_TO_MHZ = 1e-6
+NS_TO_US = 1e-3
+
+
+def convert_qubit(qubit: Qubit) -> rfsoc.Qubit:
+    """Convert `qibolab.platforms.abstract.Qubit` to `qibosoq.abstract.Qubit`"""
+    if qubit.flux:
+        return rfsoc.Qubit(qubit.flux.bias, qubit.flux.ports[0][1])
+    return rfsoc.Qubit(0.0, None)
+
+
+def set_rfsoc_pulse_shape(pulse: Pulse, rfsoc_pulse: rfsoc.Pulse):
+    """Set pulse shape parameters in rfsoc pulse object"""
+
+    rfsoc_pulse.shape = pulse.shape.name.lower()
+    if rfsoc_pulse.shape in {"gaussian", "drag"}:
+        rfsoc_pulse.rel_sigma = pulse.shape.rel_sigma
+        if rfsoc_pulse.shape == "drag":
+            rfsoc_pulse.beta = pulse.shape.beta
+
+
+def convert_pulse(pulse: Pulse, qubits: Dict[int, Qubit]) -> rfsoc.Pulse:
+    """Convert `qibolab.pulses.pulse` to `qibosoq.abstract.Pulse`"""
+    pulse_type = pulse.type.name.lower()
+    dac = getattr(qubits[pulse.qubit], pulse_type).ports[0][1]
+    adc = qubits[pulse.qubit].feedback.ports[0][1] if pulse_type == "readout" else None
+
+    try:
+        lo_frequency = getattr(qubits[pulse.qubit], pulse_type).local_oscillator._frequency
+    except NotImplementedError:
+        lo_frequency = 0
+
+    rfsoc_pulse = rfsoc.Pulse(
+        frequency=(pulse.frequency - lo_frequency) * HZ_TO_MHZ,
+        amplitude=pulse.amplitude,
+        relative_phase=np.degrees(pulse.relative_phase),
+        start=pulse.start * NS_TO_US,
+        duration=pulse.duration * NS_TO_US,
+        dac=dac,
+        adc=adc,
+        shape=None,
+        name=pulse.serial,
+        type=pulse_type,
+    )
+    set_rfsoc_pulse_shape(pulse, rfsoc_pulse)
+    return rfsoc_pulse
```

Comment:
If you have a returning function, you could simply do:
```suggestion
    return rfsoc_pulse_replace_shape(pulse, rfsoc_pulse)
```